spec/index.bs

```diff
@@ -7960,6 +7960,28 @@ dictionary GPURenderPassColorAttachment {
         render pass. If not [=map/exist|provided=] defaults to `{r: 0, g: 0, b: 0, a: 0}`. Ignored
         if {{GPURenderPassColorAttachment/loadOp}} is not {{GPULoadOp/"clear"}}.
 
+        The members of {{GPURenderPassColorAttachment/clearValue}} are all double values, so
+        they will first be converted to the fully qualified format type of
+        {{GPURenderPassColorAttachment/view}} before being set as the clear value of
+        {{GPURenderPassColorAttachment/view}}.
+
+        Let |colorAttachmentFormat| be
+        {{GPURenderPassColorAttachment/view}}.{{GPUTextureView/[[descriptor]]}}.{{GPUTextureViewDescriptor/format}}.
+        A |colorAttachmentFormat| has up to four |channel|s: `r`, `g`, `b`, and `a` (each |channel|
+        contains one scalar), normally corresponding to the concepts of red, green, blue, and alpha
+        channels of {{GPURenderPassColorAttachment/clearValue}}.
+
+        For each |channel| of |colorAttachmentFormat|:
+        - If the |channel| contains a {{float}}, signed or unsigned normalized value, the
+            corresponding member of {{GPURenderPassColorAttachment/clearValue}} will be converted
+            into the format type of |channel|.
```

Comment:
This is still underspecified:
- Still need to specify that normalized values are clamped and how "halfway" values are rounded.
- Conversion to `{{float}}` (IDL float - actually should be `{{unrestricted float}}`) isn't enough because that's just float32, but we have float16, ufloat10, and ufloat11 as well. It's OK if, for now, we just say something like "Issue: Define our own conversions for float16/ufloat10/ufloat11 following the [same pattern](https://webidl.spec.whatwg.org/#es-unrestricted-float) as {{unrestricted float}}".

Reply:
> Still need to specify that normalized values are clamped and how "halfway" values are rounded.

Oh in [D3D SPEC](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#FLOATtoUNORM) there is also no definition of "how 'halfway' values are rounded", and I remember it is a known issue that float-to-unorm-conversion is not always precise (e.g. 0.5 may be converted to either 127 or 128).

In addition, specifying a float32 value as the clear value of float/snorm/unorm formats is well supported on all the backends, and here I think we can just focus on how to handle int/uint formats.

Reply:
We don't need D3D to define how halfway values are rounded. We could spec it tightly and implementations can make it work:
- receive js double 0.5
- webgpu implementation knows the format is unorm8 and deterministically chooses a value of 128 (for example)
- webgpu implementation computes `128f / 255f` and passes that as the D3D clear value

That said, now that I'm thinking about it, this would be a lot stricter than the conversion when a shader writes an f32 to a texture. So there's not much reason to define it very strictly, as much as I want to. (We can always tighten it later.)

It also occurs to me that well-defining conversions is easier said than done for a few cases. In particular the `double` -> `f32` conversion defined by WebIDL is stricter than [the C definition](https://stackoverflow.com/q/65610811) of double->float cast, and if we define all our conversions in the WebIDL style they'll be stricter than shader f32->texture writes.

So now I'm thinking I want to propose we define the conversions here in the shader style. Unfortunately I don't know what that means exactly, but it at least means that 0.5 -> 127 and 128 are both valid. Not sure about f64->f32/f16/uf10/uf11. (Is truncation valid, as it seems to be in C? Or do you have to round to the closest value when possible?)

Reply:
Editors agreed looser conversions are fine, up to whatever we allow when converting shader f32 into a render attachment... which we don't specify yet\*. But we can define this in terms of that.

Note that conversions to integer formats are still fully strict: AFAICT in both C and WebIDL they always truncate, so there's no ambiguity about what integer you get from a double. We only have to define the behavior (clamp) when the double is out of range of the target int type.

\* We know it's not super strict (e.g. 0.5 can be either 127 or 128), but it's also probably not quite as loose as C: at least in D3D11, there's this stipulation on "Data Invertability":
https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#19.1.2%20Data%20Invertability
which says that going from memory -> shader -> memory won't change the value (except for snorm `-1.0`)

Reply:
> Note that conversions to integer formats are still fully strict: AFAICT in both C and WebIDL they always truncate, so there's no ambiguity about what integer you get from a double.

Oh yeah, IDL defines clamping-casts to int as rounding, not truncating...

Reply:
I'll propose making integers use a _non_ clamping cast, like conversion to IDL `long` (rather than `[EnforceRange] long` or `[Clamp] long`.

Reply:
Revised accordingly.